### PR TITLE
Fix error message when type argument arity is wrong

### LINKED ISSUE #28053: Expected 0-2 type arguments, but got 1
**TypeScript Version:**  3.2.0-dev.20181020

**Code**

```ts
new Map<string>();
```
(as of #28052)

**Expected behavior:**

```
No overload expects 1 type argument, but overloads do exist that expect either 0 or 2 type arguments.
```

**Actual behavior:**

```Expected 0-2 type arguments, but got 1.```

**Related Issues:** #19220

### PR DESCRIPTION
**Pull request by pesca**
Fixes #28053 

* I decided to forgo addressing the plurals issue to avoid polluting `Diagnostics.json` with all the different permutations of the error messages. I believe that should ideally be handled in the "diagnostic emitter".
* Generic functions with no overloads are handled as a special case and uses the old error message.

### CODE REVIEW

**Comment by DanielRosenwasser** [comment]
Looks good; can you resolve the merge conflict, and also make it `argument(s)`?

**Comment by pesca** [comment]
@DanielRosenwasser Merge conflict resolved.

File: src/compiler/diagnosticMessages.json [hunk 1]
```diff
@@ -2012,7 +2012,7 @@
         "category": "Error",
         "code": 2557
     },
-    "Expected {0} type arguments, but got {1}.": {
+    "Expected {0} type argument(s), but got {1}.": {
```
Comment (sandersn) [inline src/compiler/diagnosticMessages.json]: This just adds visual noise. Please revert it.

File: src/compiler/checker.ts [hunk 2]
```diff
@@ -19789,14 +19789,31 @@ namespace ts {
...
+                else if (max < argCount) {
+                    belowArgCount = Math.max(belowArgCount, max);
+                }
+            }
+            if (belowArgCount !== -Infinity && aboveArgCount !== Infinity) {
+                return createDiagnosticForNodeArray(getSourceFileOfNode(node), typeArguments, Diagnostics.No_overload_expects_0_type_argument_s_but_overloads_do_exist_that_expect_either_1_or_2_type_argument_s, argCount, belowArgCount, aboveArgCount);
             }
-            const paramCount = min === max ? min : min + "-" + max;
-            return createDiagnosticForNodeArray(getSourceFileOfNode(node), typeArguments, Diagnostics.Expected_0_type_arguments_but_got_1, paramCount, typeArguments.length);
+            return createDiagnosticForNodeArray(getSourceFileOfNode(node), typeArguments, Diagnostics.No_overload_expects_0_type_argument_s_but_an_overload_does_exist_that_expects_1_type_argument_s, argCount, belowArgCount === -Infinity ? aboveArgCount : belowArgCount);
```
Comment (sandersn) [inline src/compiler/checker.ts]: Can you revert this to the previous message? The only reason I see for the new message is that before, the expected was either min or min&ndash;max. Now expected is either above or below, but above is the common case anyway, and means basically the same as before.

File: src/compiler/diagnosticMessages.json [hunk 3]
```diff
@@ -2529,6 +2529,14 @@
         "category": "Error",
         "code": 2742
     },
+    "No overload expects {0} type argument(s), but an overload does exist that expects {1} type argument(s).": {
+        "category": "Error",
+        "code": 2743
+    },
+    "No overload expects {0} type argument(s), but overloads do exist that expect either {1} or {2} type argument(s).": {
```
Comment (sandersn) [inline src/compiler/diagnosticMessages.json]: This wording is a lot easier to scan: "Expected at least {0} or at most {1} type arguments, but got {2}."

It overpromises a bit &mdash; "at least {0}" arguments is not actually guaranteed to accept less than {0} arguments &mdash; but it's much easier to read. If the programmer overshoots, then they will get a follow-up message correcting them again.

**Comment by sandersn** [comment]
Follow up: I discovered that this is a follow-up to #24031, which does the same thing for parameters instead of type parameters. I don't think that was a good change either, but a lot of people upvoted the bug. I think the reason fewer people upvoted this one is that it's much more rare with type parameters.

Anyway, consistency is one good argument for merging this since we already have this error for parameters. That means my last comment is moot because that error message is the same as the parameter one.

File: src/compiler/diagnosticMessages.json (hunk 3 above)
Comment (sandersn) [reply src/compiler/diagnosticMessages.json]: Never mind. See below.

File: src/compiler/checker.ts (hunk 2 above)
Comment (pesca) [reply src/compiler/checker.ts]: This message is consistent with the other new message when there are overloads, but I suppose the old message is fine as well.

**Comment by pesca** [comment]
Requested changes made.

**Comment by DanielRosenwasser** [comment]
@fatcerberus yup, the error messages have been corrected and now give a better idea of what went wrong.

**Pull request merged**